### PR TITLE
User locations

### PR DESCRIPTION
**Pull request by aapeliv**
Moves to PostGIS from postgres and implements user locations.

For now this introduces the database fields for user locations, and has a basic screen for selecting one's own location.

### CODE REVIEW

File: app/frontend/src/views/EditLocation.vue [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        const selected = this.osmResponses[to]
+
+        this.center = latLng(selected.lat, selected.lon)
+        this.trueLocation = latLng(selected.lat, selected.lon)
+        // give them a random location within random angle and distance randomly between 50m - 500m from current location
+        // this is not truely uniformly random in the circle, but close enough
+        const angle = Math.random() * 2 *3.1415
+        const distance = Math.random() * 450 + 50
+        // see https://gis.stackexchange.com/a/2964
+        // 111111 m ~ 1 degree
```
Comment (kalvdans) [inline app/frontend/src/views/EditLocation.vue]: Wow, talk about magic numbers :D

File: app/pb/api.proto [hunk 2]
```diff
@@ -164,37 +164,44 @@ message User {
...
-
-  string avatar_url = 32;
+
+  // doubles are enough to describe lat/lng
+  // in EPSG4326
+  double lat = 5;
+  double lng = 6;
+  double radius = 7;
+
+  double verification = 8;
```
Comment (kalvdans) [inline app/pb/api.proto]: Why did you change the number of existing protocol entries? It is not backwards compatible, and makes the diff unnecessary long.

File: app/pb/api.proto [hunk 3]
```diff
@@ -164,37 +164,44 @@ message User {
...
-  google.protobuf.StringValue area = 30;
-  google.protobuf.StringValue house_rules = 31;
-
-  string avatar_url = 32;
+
+  // doubles are enough to describe lat/lng
+  // in EPSG4326
+  double lat = 5;
+  double lng = 6;
+  double radius = 7;
```
Comment (kalvdans) [inline app/pb/api.proto]: Please mention the units of these three doubles. (degrees, degrees, km?)

File: app/pb/api.proto (hunk 2 above)
Comment (aapeliv) [reply app/pb/api.proto]: Fixed

File: app/backend/src/couchers/models.py [hunk 4]
```diff
@@ -112,7 +123,21 @@ class User(Base):
...
+
+    @property
+    def is_missing_location(self):
+        return not self.geom or not self.geom_radius
+
+    @property
+    def coordinates(self):
+        # returns (lat, lng)
+        # we put people without coords on null island
+        # https://en.wikipedia.org/wiki/Null_Island
```
Comment (kalvdans) [inline app/backend/src/couchers/models.py:143]: Haha, cool :)

There's an exclusion zone around the weather buoy so no one can be there 
http://www.thegreenwichmeridian.org/tgm/location.php

File: app/backend/src/couchers/servicers/auth.py [hunk 5]
```diff
@@ -3,7 +3,7 @@
 from typing import Union
 
 import grpc
-from sqlalchemy import func
+from sqlalchemy.sql import func
```
Comment (kalvdans) [inline app/backend/src/couchers/servicers/auth.py:7]: Unreated changes, please move to a separate PR.

File: app/pb/api.proto [hunk 6]
```diff
@@ -214,19 +221,24 @@ message UpdateProfileReq {
   // These StringValue types allow us to determine if the value is absent versus ""
 
   google.protobuf.StringValue name = 1;
+
   google.protobuf.StringValue city = 2;
+  google.protobuf.DoubleValue lat = 22;
+  google.protobuf.DoubleValue lng = 23;
+  google.protobuf.DoubleValue radius = 24;
```
Comment (kalvdans) [inline app/pb/api.proto]: please describe (again) the unit of lat,lng and radius

File: app/pb/api.proto [hunk 7]
```diff
@@ -166,6 +166,13 @@ message User {
   string username = 2;
   string name = 3;
   string city = 4;
+
+  // doubles are enough to describe lat/lng
+  // in EPSG4326, lat & lon are in degress, check the EPS4326 definition for details
+  double lat = 33;
+  double lng = 34;
+  double radius = 35; // meters
```
Comment (kalvdans) [inline app/pb/api.proto:175]: please mention that these are returned as 0.0 when not set by the user.

File: app/pb/jail.proto [hunk 8]
```diff
@@ -26,14 +26,27 @@ service Jail {
   rpc AcceptTOS(AcceptTOSReq) returns (JailInfoRes) {
     // Accept to TOS
   }
+
+  rpc SetLocation(SetLocationReq) returns (JailInfoRes) {
+    // Set the user's location
+    // If you're not jailed, do this with api.UpdateProfile!
```
Comment (kalvdans) [inline app/pb/jail.proto:32]: Complicated with two calls... but documented so OK :)

**Review by kalvdans — CHANGES_REQUESTED** [review]
Will test run postgis tonight :)

**Review by lucaslcode — DISMISSED** [review]

**Comment by aapeliv** [comment]
@kalvdans: I commented the APIs a bit more. I don't think the minor cleanup with importing `sqlalchemy.sql.func` instead of `sqlalchemy.func` is worth spending the time separating from this and moving to an extra PR?

**Comment by kalvdans** [comment]
> @kalvdans: I commented the APIs a bit more. I don't think the minor cleanup with importing `sqlalchemy.sql.func` instead of `sqlalchemy.func` is worth spending the time separating from this and moving to an extra PR?

Is is ok with a drive-by fix. In an ideal world, it would be more effort to write a long comment than doing a small PR 👍

**Comment by kalvdans** [comment]
> Will test run postgis tonight :)

Done, it was a beast to install on Debian since it dragged in the raster support as well. All tests passes (ran without docker).

**Review by kalvdans — DISMISSED** [review]

**Comment by kalvdans** [comment]
I think you have to run "CREATE EXTENSION postgis;" when migrating an older database. Not tested it though, just reading the postgis manual.

**Comment by aapeliv** [comment]
> I think you have to run "CREATE EXTENSION postgis;" when migrating an older database. Not tested it though, just reading the postgis manual.

Yes, normally. But the docker image runs it automatically on startup

**Review by kalvdans — DISMISSED** [review]
backend still looks good after rebase, can't comment on frontend.

Tests run awfully slow though :(
* this branch: 186 passed in 457.64s (0:07:37)
* develop: 182 passed in 86.52s (0:01:26)
* develop + `CREATE EXTENSION postgis;`: 182 passed in 359.34s (0:05:59)

**Comment by aapeliv** [comment]
> Tests run awfully slow though :(
> 
>     * this branch: 186 passed in 457.64s (0:07:37)
> 
>     * develop: 182 passed in 86.52s (0:01:26)

Yes, it's because of the `CREATE EXTENSION postgis` that needs to be done after `DROP SCHEMA`.

Should come up with something faster!

**Comment by kalvdans** [comment]
> Should come up with something faster!

I tried to put the externsion in a different schema. It was faster but sadly didn't work :D
```
    engine.execute("DROP SCHEMA public CASCADE; CREATE SCHEMA IF NOT EXISTS foo; CREATE EXTENSION IF NOT EXISTS postgis WITH SCHEMA foo; CREATE SCHEMA public;")
```

**Comment by aapeliv** [comment]
We could run the whole test in a transaction and then `ROLLBACK`......

**Comment by kalvdans** [comment]
I also tried
```
Base.metadata.drop_all(engine)
engine.execute("DROP TABLE IF EXISTS alembic_version")
```
But get an error about types redefined...

**Comment by aapeliv** [comment]
We could just run each test in a different schema... And drop all at the very end...

Actually, on a further read, postgis is also installed in a schema.

Maybe we can prefix all objects with the test name...

**Comment by aapeliv** [comment]
> I also tried
> 
> ```
> Base.metadata.drop_all(engine)
> engine.execute("DROP TABLE IF EXISTS alembic_version")
> ```
> 
> But get an error about types redefined...

I think this should be the way to go. I'll have a look at the type redefinitions

**Comment by aapeliv** [comment]
I've pushed a function that allows downgrading migrations to this branch. If the migrations were correct, this should work:

```py

    if request.param == "migrations":
        # rebuild it with alembic migrations (and tear down afterwards)
        apply_migrations("upgrade")
        yield sessionmaker(bind=engine)
        apply_migrations("downgrade")
    else:
        # create everything from the current models, not incrementally through migrations
        Base.metadata.create_all(engine)
        yield sessionmaker(bind=engine)
        Base.metadata.drop_all(engine)
```

However, the migrations are not working properly. This is partly alembic, partly because we modified some settings (such as the constraint naming) halfway through.

I feel the migrations should maybe be recreated, and from now on, created and tested with some more care, to make sure the `downgrade()` also works. But that might be a longer term solution.

**Comment by kalvdans** [comment]
> I feel the migrations should maybe be recreated, and from now on, created and tested with some more care, to make sure the `downgrade()` also works. But that might be a longer term solution.

Should we revert the downgrade in this branch, merge it, and then work on speeding up tests and fixing migrations in another branch?

**Comment by aapeliv** [comment]
That might be the way to go.

**Comment by aapeliv** [comment]
@kalvdans: could you approve if you're happy for me to merge this? The migrations should work upwards, and we should fix the slow tests and downgrading migrations in another PR. This is starting to be a bit of a pain now with multiple branches/PRs off this one.

**Review by kalvdans — APPROVED** [review]
I have reverted a9dcd4b27b5676542e28d8d3b08e73e58d310bd1 and is now happy with the backend part :)

**Comment by aapeliv** [comment]
Thanks!
